src/Storages/NATS/NATSSource.cpp

```diff
@@ -0,0 +1,132 @@
+#include <Storages/NATS/NATSSource.h>
+
+#include <Formats/FormatFactory.h>
+#include <Interpreters/Context.h>
+#include <Processors/Executors/StreamingFormatExecutor.h>
+#include <Storages/NATS/ReadBufferFromNATSConsumer.h>
+
+namespace DB
+{
+
+static std::pair<Block, Block> getHeaders(const StorageSnapshotPtr & storage_snapshot)
+{
+    auto non_virtual_header = storage_snapshot->metadata->getSampleBlockNonMaterialized();
+    auto virtual_header = storage_snapshot->getSampleBlockForColumns({"_subject"});
+
+    return {non_virtual_header, virtual_header};
+}
+
+static Block getSampleBlock(const Block & non_virtual_header, const Block & virtual_header)
+{
+    auto header = non_virtual_header;
+    for (const auto & column : virtual_header)
+        header.insert(column);
+
+    return header;
+}
+
+NATSSource::NATSSource(
+    StorageNATS & storage_,
+    const StorageSnapshotPtr & storage_snapshot_,
+    ContextPtr context_,
+    const Names & columns,
+    size_t max_block_size_)
+    : NATSSource(storage_, storage_snapshot_, getHeaders(storage_snapshot_), context_, columns, max_block_size_)
+{
+}
+
+NATSSource::NATSSource(
+    StorageNATS & storage_,
+    const StorageSnapshotPtr & storage_snapshot_,
+    std::pair<Block, Block> headers,
+    ContextPtr context_,
+    const Names & columns,
+    size_t max_block_size_)
+    : SourceWithProgress(getSampleBlock(headers.first, headers.second))
+    , storage(storage_)
+    , storage_snapshot(storage_snapshot_)
+    , context(context_)
+    , column_names(columns)
+    , max_block_size(max_block_size_)
+    , non_virtual_header(std::move(headers.first))
+    , virtual_header(std::move(headers.second))
+{
+    storage.incrementReader();
+}
+
+
+NATSSource::~NATSSource()
+{
+    storage.decrementReader();
+
+    if (!buffer)
+        return;
+
+    storage.pushReadBuffer(buffer);
+}
+
+Chunk NATSSource::generate()
+{
+    auto chunk = generateImpl();
+
+    return chunk;
+}
```

Comment:
if there is nothing else apart from calling just `generateImpl` then maybe get rid of impl?